tracing-subscriber/src/filter/subscriber_filters/combinator.rs

```diff
@@ -393,7 +403,12 @@ where
     /// If the wrapped filter would enable a span or event, it will be disabled. If
     /// it would disable a span or event, that span or event will be enabled.
     ///
+    /// This inversion only considers the [`enabled`] filter;
+    /// it does not call the [`event_enabled`] filter.
+    ///
     /// [`Filter`]: crate::subscribe::Filter
+    /// [`enabled`]: crate::subscribe::Filter::enabled
+    /// [`event_enabled`]: crate::subscribe::Filter::event_enabled
```

Comment:
This isn't quite right; it also inverts `callsite_enabled`. Maybe:
```suggestion
    /// This inverts the values returned by the [`enabled`] and [`callsite_enabled`]
    /// methods on the wrapped filter; it does *not* invert [`event_enabled`], as
    /// implementing that method is optional, and filters which do not implement
    /// filtering on event field values will return `true` even for events that their
    /// [`enabled`] method would disable.
    ///
    /// [`Filter`]: crate::subscribe::Filter
    /// [`enabled`]: crate::subscribe::Filter::enabled
    /// [`event_enabled`]: crate::subscribe::Filter::event_enabled
    /// [`callsite_enabled`]: crate::subscribe::Filter::callsite_enabled
```

Reply:
in retrospect...maybe `event_enabled` should have been an `Option<bool>` to distinguish between "i am actively enabling this" and "i don't implement filtering beyond normal metadata-based `enabled`", so we could invert the value inside the `Option`, but not touch filters that just ignore that method. But, on the other hand, that's significantly more complex, and I'm not sure how useful the `Not` combinator is anyway...

Reply:
`event_enabled -> Option<bool>` isn't necessary for `Not` having "fully proper" semantics. `!(callsite_enabled && enabled && event_enabled)` distributes into `!callsite_enabled || !enabled || !event_enabled`. So the *fully proper* implementation of `not` would be IIUC roughly

```rust
fn callsite_enabled() =
    match inner.callsite_enabled():
        Never => Always,
        Always => Never,
        Sometimes => Sometimes,

fn enabled() =
    callsite_enabled() == Always || !inner.enabled() || is_event()

fn event_enabled() =
    callsite_enabled() == Always || !inner.enabled() || !inner.event_enabled()
```

(but with caching to only call impure `inner` once.) Working-out:

```rust
= match callsite_enabled():
    Always => true,
    Never => false,
    Sometimes => enabled() && (is_span() || event_enabled())

= inner.callsite_enabled().match {
    Never => Always,
    Always => Never,
    Sometimes => Sometimes,
}.match {
    Always => true,
    Sometimes => (callsite_enabled() == Always || !inner.enabled() || is_event()) &&
        (is_span() || (callsite_enabled() == Always || !inner.enabled() || !inner.event_enabled()))
}

= match inner.callsite_enabled():
    Never => true,
    Always => false,
    Sometimes => match:
        is_span() => !inner.enabled(),
        is_event() => !inner.enabled() || !inner.event_enabled()

= !inner.callsite_enabled() || !inner.enabled() || !inner.event_enabled()
= !(inner.callsite_enabled() && inner.enabled() && inner.event_enabled())
where
    Always && _ = Always
    Always || _ = Always
    Never && _ = Never
    Never || _ = Never

    Sometimes && true = Sometimes
    Sometimes && false = Never
    Sometimes || true = Sometimes
    Sometimes || false = Never

    !Always = Never
    !Never = Always
    !Sometimes = Sometimes

    (Always | Sometimes) as bool = true
    Never as bool = false
```

What's *currently implemented* is IIUC `!callsite_enabled && !enabled`, which is different 🙂 but perhaps *marginally* more likely to be useful?